spring-integration-core/src/main/java/org/springframework/integration/util/CloseableIterator.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.util;
+
+import java.io.Closeable;
+import java.util.Iterator;
+
+/**
+ * A {@link CloseableIterator} is intended to be used when it may hold resources (such as file or socket handles).
+ * This allows implementations to clean up any resources they need to keep open to iterate over elements.
+ *
+ * @author Ruslan Stelmachenko
+ *
+ * @since 4.3.15, 5.0.4
+ */
+public interface CloseableIterator<E> extends Iterator<E>, Closeable {
+
+	/**
+	 * Closes this iterator and releases any system resources associated
+	 * with it. If the iterator is already closed then invoking this
+	 * method has no effect.
```

Comment:
I would say that this second sentence with the `if` is too bold: we can't guarantee the target implementation.
May be just remove it altogether?

Reply:
Actually, this sentence is from [Closeable](https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html#close--) interface. I think it is more like a contract which all implementors should follow. But I agree that, in this formulation it is too vague. Maybe just remove it is better.